README.md
# Sensu SNMP Trap Extension

A SNMP trap listener for the Sensu client process. This check
extension creates a SNMPv2 trap listener, loads & imports MIB files,
and attempts to translate SNMP traps into Sensu check results. Sensu
proxy clients are created for SNMP devices, using the device hostname
as the client name, when a SNMP trap is received.

The SNMP trap to Sensu check result translation logic is currently
being improved. Failure to translate the SNMP trap will produce a
check result (and event) like the following (addresses have been
redacted):

![screenshot](https://raw.github.com/sensu-extensions/sensu-extensions-snmp-trap/master/iflinkdown.png)

## Inspiration

This SNMP trap extension was inspired by the work done by Toby Jackson
on [SNMPTrapHandler](https://github.com/warmfusion/sensu-extension-snmptrap).

## Installation

This extension requires Sensu version >= 0.26.

On a Sensu client machine.

```
sensu-install -e snmp-trap:0.0.33
```

Edit `/etc/sensu/conf.d/extensions.json` to load it.

``` json
{
  "extensions": {
    "snmp-trap": {
      "version": "0.0.33"
    }
  }
}
```

## Configuration

Edit `/etc/sensu/conf.d/snmp_trap.json` to change its configuration.

``` json
{
  "snmp_trap": {
    "community": "secret",
    "result_attributes": {
      "datacenter": "DC01"
    },
    "result_status_map": [
      ["authenticationFailure", 0]
    ],
    "client_socket_bind": "192.168.0.3",
    "client_socket_port": 1234
  }
}
```

|attribute|type|default|description|
|----|----|----|---|
|bind|string|0.0.0.0|IP to bind the SNMP trap listener to|
|port|integer|1062|Port to bind the SNMP trap listener to|
|community|string|"public"|SNMP community string to use|
|mibs_dir|string|"/etc/sensu/mibs"|MIBs directory to import and load MIBs from|
|imported_dir|string|"$TMPDIR/sensu_snmp_imported_mibs"|Directory to store imported MIB data in|
|handlers|array|["default"]|Handlers to specify in Sensu check results|
|result_attributes|hash|{}|Custom check result attributes to add to every SNMP trap Sensu check result|
|result_map|array|[]|SNMP trap varbind to Sensu check result translation mappings|
|result_status_map|array|[]|SNMP trap varbind to Sensu check result status mappings|
|client_socket_bind|string|"127.0.0.1"|IP to send events to when handled|
|client_socket_port|integer|3030|Port to send events to when handled|

### Result Map Examples

The configurable result map allows you to define SNMP trap varbind to
Sensu check result attribute mappings. A mapping is comprised of a
varbind name regular expression and a check attribute. For example, if
you expect SNMP traps with a varbind name that contains
"AlertDescription" and you would like to use its value as the Sensu
check result output:

``` json
{
  "snmp_trap": {
    "...": "...",
    "result_map": [
      ["/description/i", "output"]
    ]
  }
}
```

Configuring a result map does not replace the built-in mappings, the
configured mappings take precedence over them.

### Result Status Map Examples

The configurable result status map allows you to define SNMP trap
varbind to numeric Sensu check result status value mappings. A
mapping is comprised of a varbind name regular expression and an check
exit status (e.g. 1-255). For example, if you expect SNMP traps with a
varbind name that contains "CriticalError" and you would like to set
the Sensu check result status to `2` (critical):

``` json
{
  "snmp_trap": {
    "...": "...",
    "result_status_map": [
      ["/critical/i", 2]
    ]
  }
}
```

Configuring a result status map does not replace the built-in
mappings, the configured mappings take precedence over them.


CHANGELOG.md
# Change Log

This project adheres to [Semantic Versioning](http://semver.org/).

This CHANGELOG follows the format listed [here](https://github.com/sensu-plugins/community/blob/master/HOW_WE_CHANGELOG.md)

## [Unreleased]
### Added
- searches the `MIB` folder recursively for files (@landondao1)

## [0.1.0] - 2018-04-06
### Added
- allow configuration of client sockets (@dudefellah)

## [0.0.33] - 2017-01-19
### Changed
- simplified map config (@portertech)

[Unreleased]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/0.2.0...HEAD
[0.2.0]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/0.1.0...0.2.0
[0.1.0]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.33...0.1.0
[0.0.33]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.32...v0.0.33
[0.0.32]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.31...v0.0.32
[0.0.31]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.30...v0.0.31
[0.0.30]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.29...v0.0.30
[0.0.29]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.28...v0.0.29
[0.0.28]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.27...v0.0.28
[0.0.27]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.26...v0.0.27
[0.0.26]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.25...v0.0.26
[0.0.25]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.24...v0.0.25
[0.0.24]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.23...v0.0.24
[0.0.23]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.22...v0.0.23
[0.0.22]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.21...v0.0.22
[0.0.21]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.20...v0.0.21
[0.0.20]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.19...v0.0.20
[0.0.19]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.18...v0.0.19
[0.0.18]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.17...v0.0.18
[0.0.17]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.16...v0.0.17
[0.0.16]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.15...v0.0.16
[0.0.15]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.14...v0.0.15
[0.0.14]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.13...v0.0.14
[0.0.13]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.12...v0.0.13
[0.0.12]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.11...v0.0.12
[0.0.11]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.10...v0.0.11
[0.0.10]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.9...v0.0.10
[0.0.9]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.8...v0.0.9
[0.0.8]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.7...v0.0.8
[0.0.7]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.6...v0.0.7
[0.0.6]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.5...v0.0.6
[0.0.5]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.4...v0.0.5
[0.0.4]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.3...v0.0.4
[0.0.3]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.2...v0.0.3
[0.0.2]: https://github.com/sensu-plugins/sensu-extensions-snmp-trap/compare/v0.0.1...v0.0.2


spec/helpers.rb
require "rspec"
require "eventmachine"
require "sensu/logger"
require "sensu/settings"
require "securerandom"

module Helpers
  def timer(delay, &callback)
    periodic_timer = EM::PeriodicTimer.new(delay) do
      callback.call
      periodic_timer.cancel
    end
  end

  def async_wrapper(&callback)
    EM.run do
      timer(10) do
        raise "test timed out"
      end
      callback.call
    end
  end

  def async_done
    EM.stop_event_loop
  end

  def logger
    Sensu::Logger.get(:log_level => :fatal)
  end

  def settings
    Sensu::Settings.get
  end

  def epoch
    Time.now.to_i
  end

  def client_template
    {
      :name => "i-424242",
      :address => "127.0.0.1",
      :subscriptions => [
        "test"
      ],
      :timestamp => epoch
    }
  end

  def check_template
    {
      :name => "test",
      :type => "standard",
      :issued => epoch,
      :command => "echo WARNING && exit 1",
      :subscribers => [
        "test"
      ],
      :interval => 1,
      :output => "WARNING",
      :status => 1,
      :executed => epoch,
      :history => [0, 1]
    }
  end

  def event_template
    one_second_ago = epoch - 1
    {
      :id => ::SecureRandom.uuid,
      :client => client_template,
      :check => check_template,
      :last_ok => one_second_ago,
      :last_state_change => one_second_ago,
      :occurrences => 1,
      :occurrences_watermark => 1,
      :action => :create
    }
  end

  class TestServer < EM::Connection
    include RSpec::Matchers

    attr_accessor :expected

    def receive_data(data)
      if @expected
        expect(Sensu::JSON.load(data)).to eq(Sensu::JSON.load(@expected))
        EM::stop_event_loop
      end
    end
  end
end


spec/snmp-trap_spec.rb
require File.join(File.dirname(__FILE__), "helpers")
require "sensu/extensions/snmp-trap"
require "socket"

describe "Sensu::Extension::SNMPTrap" do
  include Helpers

  before do
    @extension = Sensu::Extension::SNMPTrap.new
    @extension.settings = {
      :snmp_trap => {
        :mibs_dir => File.join(File.dirname(__FILE__), "mibs")
      }
    }
    @extension.logger = Sensu::Logger.get(:log_level => :fatal)
  end

  let(:snmpv2_pdu) do
    varbind_list = [
      SNMP::VarBind.new(SNMP::SYS_UP_TIME_OID, SNMP::TimeTicks.new(20)),
      SNMP::VarBind.new(SNMP::SNMP_TRAP_OID_OID, SNMP::ObjectId.new("1.3.6.1.4.1.45717.1.0")),
      SNMP::VarBind.new("1.3.6.1.4.1.45717.1.1.1.2", SNMP::OctetString.new("alert")),
      SNMP::VarBind.new("1.3.6.1.4.1.45717.1.1.1.5", SNMP::Integer32.new(2))
    ]
    SNMP::SNMPv2_Trap.new(1, SNMP::VarBindList.new(varbind_list))
  end

  let(:snmpv2_message) do
    SNMP::Message.new(:SNMPv2c, "public", snmpv2_pdu).encode
  end

  it "can run" do
    async_wrapper do
      EM::open_datagram_socket("127.0.0.1", 3030, Helpers::TestServer) do |server|
        server.expected = '{"source": "localhost", "name": "SENSU-ENTERPRISE-NOTIFY-MIB--sensuEnterpriseEventTrap", "output": "alert", "status": 2, "handlers": ["default"], "snmp_trap": {"DISMAN-EXPRESSION-MIB::sysUpTimeInstance": 20, "SNMPv2-MIB::snmpTrapOID.0": "SNMPv2-SMI::enterprises.45717.1.0", "SENSU-ENTERPRISE-NOTIFY-MIB::sensuNotification": "alert", "SENSU-ENTERPRISE-NOTIFY-MIB::sensuCheckSeverity": 2}}'
      end
      EM.next_tick do
        EM::open_datagram_socket("0.0.0.0", 0, nil) do |socket|
          socket.send_datagram(snmpv2_message, "127.0.0.1", 1062)
          socket.close_connection_after_writing
        end
      end
    end
  end

  it "can load mibs from files recursively" do
    @extension.create_mibs_map!
    mibs = @extension.instance_variable_get(:@mibs_map)
    expect(mibs).to have_key("APACHE-MIB")
  end
end


lib/sensu/extensions/snmp-trap/snmp-patch.rb
require "snmp"

module SNMP
  class MIB
    def warn(message)
      @sensu_logger ||= Sensu::Logger.get
      @sensu_logger.warn("snmp trap check extension mib warning", :warning => message.downcase)
    end

    class << self
      def warn(message)
        @sensu_logger ||= Sensu::Logger.get
        @sensu_logger.warn("snmp trap check extension mib warning", :warning => message.downcase)
      end

      def import_module(module_file, mib_dir=DEFAULT_MIB_PATH, raw_arguments=nil)
        raise "smidump tool must be installed" unless import_supported?
        FileUtils.makedirs mib_dir
        # PATCH: add arguments support
        arguments = raw_arguments ? " #{raw_arguments}" : ""
        # PATCH: redirect STDERR to /dev/null
        mib_hash = `smidump#{arguments} -k -f python #{module_file} 2>/dev/null`
        mib = eval_mib_data(mib_hash)
        if mib
          module_name = mib["moduleName"]
          raise "#{module_file}: invalid file format; no module name" unless module_name
          if mib["nodes"]
            oid_hash = {}
            mib["nodes"].each { |key, value| oid_hash[key] = value["oid"] }
            if mib["notifications"]
              mib["notifications"].each { |key, value| oid_hash[key] = value["oid"] }
            end
            File.open(module_file_name(module_name, mib_dir), 'w') do |file|
              YAML.dump(oid_hash, file)
              file.puts
            end
            module_name
          else
            # PATCH: downcase and removed ***
            warn "no nodes defined in: #{module_file}"
            nil
          end
        else
          # PATCH: downcase and removed ***
          warn "import failed for: #{module_file}"
          nil
        end
      end
    end
  end
end


lib/sensu/extensions/snmp-trap.rb
require "sensu/extension"
require "sensu/extensions/snmp-trap/snmp-patch"
require "thread"

module Sensu
  module Extension
    class SNMPTrap < Check

      RESULT_MAP = [
        [/checkname/i, :name],
        [/notification/i, :output],
        [/description/i, :output],
        [/pansystemseverity/i, Proc.new { |value| value > 3 ? 2 : 0 }, :status],
        [/severity/i, :status]
      ]

      RESULT_STATUS_MAP = [
        [/down/i, 2],
        [/authenticationfailure/i, 1]
      ]

      RUBY_ASN1_MAP = {
        "INTEGER" => :to_i,
        "OCTET STRING" => :to_s,
        "OBJECT IDENTIFIER" => :to_s,
        "IpAddress" => :to_s,
        "Counter32" => :to_i,
        "Gauge32" => :to_i,
        "Unsigned32" => :to_i,
        "TimeTicks" => :to_i,
        "Opaque" => :to_s,
        "Counter64" => :to_i
      }

      def name
        "snmp_trap"
      end

      def description
        "receives snmp traps and translates them to check results"
      end

      def definition
        {
          :name => name,
          :publish => false
        }
      end

      def options
        return @options if @options
        @options = {
          :bind => "0.0.0.0",
          :port => 1062,
          :community => "public",
          :mibs_dir => "/etc/sensu/mibs",
          :imported_dir => File.join(Dir.tmpdir, "sensu_snmp_imported_mibs"),
          :handlers => ["default"],
          :result_attributes => {},
          :result_map => [],
          :result_status_map => [],
          :client_socket_bind => "127.0.0.1",
          :client_socket_port => 3030
        }
        @options.merge!(@settings[:snmp_trap]) if @settings[:snmp_trap].is_a?(Hash)
        @options
      end

      def convert_to_map(configured_map)
        configured_map.map do |mapping|
          [
            Regexp.new(mapping.first),
            mapping.last.is_a?(String) ? mapping.last.to_sym : mapping.last
          ]
        end
      end

      def result_map
        return @result_map if @result_map
        if options[:result_map] && options[:result_map].is_a?(Array)
          @result_map = convert_to_map(options[:result_map]) + RESULT_MAP
        else
          @result_map = RESULT_MAP
        end
      end

      def result_status_map
        return @result_status_map if @result_status_map
        if options[:result_status_map] && options[:result_status_map].is_a?(Array)
          @result_status_map = convert_to_map(options[:result_status_map]) + RESULT_STATUS_MAP
        else
          @result_status_map = RESULT_STATUS_MAP
        end
      end

      def start_snmpv2_listener!
        @listener = SNMP::TrapListener.new(
          :host => options[:bind],
          :port => options[:port],
          :community => options[:community]) do |listener|
          listener.on_trap_v2c do |trap|
            @logger.debug("snmp trap check extension received a v2 trap")
            @traps << trap
          end
        end
      end

      def determine_mib_preload(module_name)
        preload = []
        if @mibs_map[module_name]
          imports = @mibs_map[module_name][:imports]
          # two enumerators are required for preload ordering
          imports.each do |import|
            if @mibs_map[import]
              preload << @mibs_map[import][:mib_file]
            end
          end
          imports.each do |import|
            preload << determine_mib_preload(import)
          end
        else
          @logger.warn("snmp trap check extension unknown mib preload module", :module_name => module_name)
        end
        preload.flatten
      end

      def create_mibs_map!
        @logger.info("snmp trap check extension creating mibs map", :mibs_dir => options[:mibs_dir])
        @mibs_map = {}
        Dir.glob(File.join(options[:mibs_dir], "**", "*")).each do |mib_file|
          begin
            mib_contents = IO.read(mib_file).force_encoding("UTF-8")
            module_name = mib_contents.scan(/([\w-]+)\s+DEFINITIONS\s+::=\s+BEGIN/).flatten.first
            details = {
              :mib_file => mib_file,
              :imports => mib_contents.scan(/FROM\s+([\w-]+)/).flatten
            }
            if @mibs_map.has_key?(module_name)
              @logger.warn("snmp trap check extension overriding mib map entry", {
                :module_name => module_name,
                :old_details => @mibs_map[module_name],
                :new_details => details
              })
            end
            @mibs_map[module_name] = details
          rescue => error
            @logger.error("snmp trap check extension mibs map error", {
              :mib_file => mib_file,
              :error => error.to_s
            })
          end
        end
        @mibs_map.each_key do |module_name|
          @mibs_map[module_name][:preload] = determine_mib_preload(module_name)
        end
        @mibs_map
      end

      def import_mibs!
        @logger.info("snmp trap check extension importing mibs", :mibs_dir => options[:mibs_dir])
        @mibs_map.each do |module_name, details|
          @logger.debug("snmp trap check extension importing mib", {
            :module_name => module_name,
            :details => details
          })
          begin
            @logger.debug("snmp trap check extension mib dependencies", {
              :module_name => module_name,
              :details => details
            })
            unless details[:preload].empty?
              arguments = "-p "
              arguments << details[:preload].map { |preload| preload }.join(" -p ")
            else
              arguments = nil
            end
            SNMP::MIB.import_module(details[:mib_file], options[:imported_dir], arguments)
          rescue StandardError, SyntaxError => error
            @logger.debug("snmp trap check extension failed to import mib", {
              :module_name => module_name,
              :details => details,
              :error => error
            })
          end
        end
      end

      def load_mibs!
        @logger.info("snmp trap check extension loading mibs", :imported_dir => options[:imported_dir])
        @mibs = SNMP::MIB.new
        @logger.debug("snmp trap check extension loading mibs")
        SNMP::MIB.list_imported(/.*/, SNMP::MIB::DEFAULT_MIB_PATH).each do |module_name|
          @logger.debug("snmp trap check extension loading mib", :module_name => module_name)
          @mibs.load_module(module_name, SNMP::MIB::DEFAULT_MIB_PATH)
        end
        SNMP::MIB.list_imported(/.*/, options[:imported_dir]).each do |module_name|
          @logger.debug("snmp trap check extension loading mib", :module_name => module_name)
          @mibs.load_module(module_name, options[:imported_dir])
        end
        @mibs
      end

      def send_result(result)
        socket = UDPSocket.new
        socket.send(Sensu::JSON.dump(result), 0, options[:client_socket_bind], options[:client_socket_port])
        socket.close
      end

      def determine_hostname(address)
        begin
          Resolv.getname(address)
        rescue Resolv::ResolvError
          @logger.debug("snmp trap check extension unable to resolve hostname", :address => address)
          address
        end
      end

      def determine_trap_oid(trap)
        varbind = trap.varbind_list.detect do |varbind|
          varbind.name.to_oid == SNMP::SNMP_TRAP_OID_OID
        end
        begin
          @mibs.name(varbind.value.to_oid).gsub(/[^\w\.-]/i, "-")
        rescue
          varbind.value.to_s.gsub(/[^\w\.-]/i, "-") rescue "trap_oid_unknown"
        end
      end

      def determine_trap_name(trap)
        oid_symbolic_name = determine_trap_oid(trap)
        if oid_symbolic_name =~ /link(down|up)/i
          name = "link_status"
          trap.varbind_list.each do |varbind|
            symbolic_name = @mibs.name(varbind.name.to_oid)
            if symbolic_name =~ /ifindex/i || symbolic_name =~ /systemobject/i
              type_conversion = RUBY_ASN1_MAP[varbind.value.asn1_type]
              if type_conversion
                value = varbind.value.send(type_conversion)
                unless value == ""
                  name = "#{name}_#{value}"
                end
              end
            end
          end
          name
        else
          oid_symbolic_name
        end
      end

      def determine_trap_output(trap)
        oid_symbolic_name = determine_trap_oid(trap)
        if matched = /link(down|up)/i.match(oid_symbolic_name)
          link_status = matched[1].downcase
          output = "link is #{link_status}"
          trap.varbind_list.each do |varbind|
            symbolic_name = @mibs.name(varbind.name.to_oid)
            if symbolic_name =~ /ifalias/i || symbolic_name =~ /ifdesc/i
              type_conversion = RUBY_ASN1_MAP[varbind.value.asn1_type]
              if type_conversion
                value = varbind.value.send(type_conversion)
                unless value == ""
                  output = "#{output} (#{value})"
                end
              end
            end
          end
          output
        else
          "received snmp trap"
        end
      end

      def determine_trap_status(trap)
        oid_symbolic_name = determine_trap_oid(trap)
        mapping = result_status_map.detect do |mapping|
          oid_symbolic_name =~ mapping.first
        end
        mapping ? mapping.last : 0
      end

      def process_trap(trap)
        @logger.debug("snmp trap check extension processing a v2 trap")
        result = options[:result_attributes].merge(
          {
            :source => determine_hostname(trap.source_ip),
            :handlers => options[:handlers],
            :snmp_trap => {}
          }
        )
        trap.varbind_list.each do |varbind|
          symbolic_name = @mibs.name(varbind.name.to_oid)
          type_conversion = RUBY_ASN1_MAP[varbind.value.asn1_type]
          if type_conversion
            value = varbind.value.send(type_conversion)
            result[:snmp_trap][symbolic_name] = value
            mapping = result_map.detect do |mapping|
              symbolic_name =~ mapping.first
            end
            if mapping && !result[mapping.last]
              if mapping.size == 3
                result[mapping.last] = mapping[1].call(value)
              else
                result[mapping.last] = value
              end
            end
          else
            @logger.error("snmp trap check extension failed to convert varbind", {
              :symbolic_name => symbolic_name,
              :asn1_type => varbind.value.asn1_type,
              :raw_value => varbind.value
            })
          end
        end
        result[:name] ||= determine_trap_name(trap)
        result[:output] ||= determine_trap_output(trap)
        result[:status] ||= determine_trap_status(trap)
        send_result(result)
      end

      def start_trap_processor!
        @processor = Thread.new do
          create_mibs_map!
          import_mibs!
          load_mibs!
          loop do
            process_trap(@traps.pop)
          end
        end
        @processor.abort_on_exception = true
        @processor
      end

      def post_init
        @traps = Queue.new
        start_snmpv2_listener!
        start_trap_processor!
      end

      def stop
        @listener.kill if @listener
        @processor.kill if @processor
      end

      def run(event, &callback)
        yield "no-op", 0
      end
    end
  end
end
